File: 125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        chars = [char.lower() for char in list(s) if char.isalnum()]
        l = 0
        r = len(chars) - 1

        while r >= l and chars[l] == chars[r]:
            r -= 1
            l += 1

        return l > r


s = Solution()
print(s.isPalindrome("A man, a plan, a canal: Panama"))
print(s.isPalindrome("race a car"))
print(s.isPalindrome("abccba"))
print(s.isPalindrome("abcba"))
print(s.isPalindrome("a"))
print(s.isPalindrome("0P"))
print(s.isPalindrome("    a    "))

File: 125-valid-palindrome.ts
namespace Problem125 {
  function isAlphaNumeric(char: string): boolean {
    return (char >= 'a' && char <= 'z') || (char >= '0' && char <= '9');
  }

  function isPalindrome(s: string): boolean {
    let l = 0;
    let r = s.length - 1;

    while (r > l) {
      const left = s[l].toLowerCase();
      const right = s[r].toLowerCase();

      if (!isAlphaNumeric(left)) {
        l++;
        continue;
      }

      if (!isAlphaNumeric(right)) {
        r--;
        continue;
      }

      if (left !== right) {
        return false;
      }

      l++;
      r--;
    }

    return true;
  }

  console.log(isPalindrome('A man, a plan, a canal: Panama'));
}
